File: src/flet_secure_storage/secure_storage.py
from dataclasses import field
from typing import Any

from flet.controls.base_control import control
from flet.controls.services.service import Service
from flet_secure_storage.options import (
    AndroidOptions,
    IOSOptions,
    MacOsOptions,
    WebOptions,
    WindowsOptions,
)


@control("SecureStorage")
class SecureStorage(Service):
    ios_options: IOSOptions = field(default_factory=lambda: IOSOptions())
    android_options: AndroidOptions = field(default_factory=lambda: AndroidOptions())
    windows_options: WindowsOptions = field(default_factory=lambda: WindowsOptions())
    macos_options: MacOsOptions = field(default_factory=lambda: MacOsOptions())
    web_options: WebOptions = field(default_factory=lambda: WebOptions())

    async def set(
        self,
        key: str,
        value: Any,
        web: WebOptions | None = None,
        ios: IOSOptions | None = None,
        macos: MacOsOptions | None = None,
        android: AndroidOptions | None = None,
        windows: WindowsOptions | None = None,
    ) -> None:
        if value is None:
            raise ValueError("value can't be None")
        return await self._invoke_method(
            method_name="set",
            arguments={
                "key": key,
                "value": value,
                "web": web,
                "ios": ios,
                "macos": macos,
                "android": android,
                "windows": windows,
            },
        )

    async def get(
        self,
        key: str,
        web: WebOptions | None = None,
        ios: IOSOptions | None = None,
        macos: MacOsOptions | None = None,
        android: AndroidOptions | None = None,
        windows: WindowsOptions | None = None,
    ) -> str | None:
        return await self._invoke_method(
            method_name="get",
            arguments={
                "key": key,
                "web": web,
                "ios": ios,
                "macos": macos,
                "android": android,
                "windows": windows,
            },
        )

    async def contains_key(
        self,
        key: str,
        web: WebOptions | None = None,
        ios: IOSOptions | None = None,
        macos: MacOsOptions | None = None,
        android: AndroidOptions | None = None,
        windows: WindowsOptions | None = None,
    ) -> bool:
        return await self._invoke_method(
            method_name="contains_key",
            arguments={
                "key": key,
                "web": web,
                "ios": ios,
                "macos": macos,
                "android": android,
                "windows": windows,
            },
        )

    async def remove(
        self,
        key: str,
        web: WebOptions | None = None,
        ios: IOSOptions | None = None,
        macos: MacOsOptions | None = None,
        android: AndroidOptions | None = None,
        windows: WindowsOptions | None = None,
    ) -> None:
        return await self._invoke_method(
            method_name="remove",
            arguments={
                "key": key,
                "web": web,
                "ios": ios,
                "macos": macos,
                "android": android,
                "windows": windows,
            },
        )

    async def get_all(
        self,
        web: WebOptions | None = None,
        ios: IOSOptions | None = None,
        macos: MacOsOptions | None = None,
        android: AndroidOptions | None = None,
        windows: WindowsOptions | None = None,
    ) -> dict[str, Any]:
        return await self._invoke_method(
            method_name="get_all",
            arguments={
                "web": web,
                "ios": ios,
                "macos": macos,
                "android": android,
                "windows": windows,
            },
        )

    async def clear(
        self,
        web: WebOptions | None = None,
        ios: IOSOptions | None = None,
        macos: MacOsOptions | None = None,
        android: AndroidOptions | None = None,
        windows: WindowsOptions | None = None,
    ) -> None:
        return await self._invoke_method(
            method_name="clear",
            arguments={
                "web": web,
                "ios": ios,
                "macos": macos,
                "android": android,
                "windows": windows,
            },
        )

File: src/flutter/flet_secure_storage/lib/src/secure_storage.dart
import 'package:flet/flet.dart';
import 'package:flutter/material.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'utils/secure_storage.dart';


class SecureStorageService extends FletService {
  SecureStorageService({required super.control});
  late FlutterSecureStorage storage;
  IOSOptions iosOptions = IOSOptions.defaultOptions;
  AndroidOptions androidOptions = AndroidOptions.defaultOptions;
  LinuxOptions linuxOptions = LinuxOptions.defaultOptions;
  WindowsOptions windowsOptions = WindowsOptions.defaultOptions;
  WebOptions webOptions = WebOptions.defaultOptions;
  MacOsOptions macOsOptions = MacOsOptions.defaultOptions;  

  @override
  void init() {
    super.init();
    debugPrint("SecureStorageService(${control.id}).init: ${control.properties}");
    control.addInvokeMethodListener(_invokeMethod);
  }

  @override
  void update() {
    debugPrint("SecureStorageService(${control.id}).update: ${control.properties}");
  }

  Future<dynamic> _invokeMethod(String name, dynamic args) async {
    storage = FlutterSecureStorage(
      lOptions: linuxOptions,
      aOptions: parseAndroidOptions(control.get<Map>("android_options"), androidOptions)!,
      wOptions: parseWindowsOptions(control.get<Map>("windows_options"), windowsOptions)!,
      mOptions: parseMacOptions(control.get<Map>("macos_options"), macOsOptions)!,
      iOptions: parseIosOptions(control.get<Map>("ios_options"), iosOptions)!,
      webOptions: parseWebOptions(control.get<Map>("web_options"), webOptions)!,
    );
    switch (name) {
      case "set":
        return await storage.write(
          key: args["key"]!,
          value: args["value"]!,
          aOptions: parseAndroidOptions(args?["android"], storage.aOptions)!,
          iOptions: parseIosOptions(args?["ios"], storage.iOptions)!,
          lOptions: storage.lOptions,
          wOptions: parseWindowsOptions(args?["windows"], storage.wOptions)!,
          webOptions: parseWebOptions(args?["web"], storage.webOptions)!,
          mOptions: parseMacOptions(args?["macos"], storage.mOptions as MacOsOptions)!,
        );
      case "get":
        return await storage.read(
          key: args["key"]!,
          aOptions: parseAndroidOptions(args?["android"], storage.aOptions)!,
          iOptions: parseIosOptions(args?["ios"], storage.iOptions)!,
          lOptions: storage.lOptions,
          wOptions: parseWindowsOptions(args?["windows"], storage.wOptions)!,
          webOptions: parseWebOptions(args?["web"], storage.webOptions)!,
          mOptions: parseMacOptions(args?["macos"], storage.mOptions as MacOsOptions)!,
        );
      case "get_all":
        return await storage.readAll(
          aOptions: parseAndroidOptions(args?["android"], storage.aOptions)!,
          iOptions: parseIosOptions(args?["ios"], storage.iOptions)!,
          lOptions: storage.lOptions,
          wOptions: parseWindowsOptions(args?["windows"], storage.wOptions)!,
          webOptions: parseWebOptions(args?["web"], storage.webOptions)!,
          mOptions: parseMacOptions(args?["macos"], storage.mOptions as MacOsOptions)!,
        );
      case "contains_key":
        return await storage.containsKey(
          key: args["key"]!,
          aOptions: parseAndroidOptions(args?["android"], storage.aOptions)!,
          iOptions: parseIosOptions(args?["ios"], storage.iOptions)!,
          lOptions: storage.lOptions,
          wOptions: parseWindowsOptions(args?["windows"], storage.wOptions)!,
          webOptions: parseWebOptions(args?["web"], storage.webOptions)!,
          mOptions: parseMacOptions(args?["macos"], storage.mOptions as MacOsOptions)!,
        );
      case "remove":
        return await storage.delete(
          key: args["key"]!,
          aOptions: parseAndroidOptions(args?["android"], storage.aOptions)!,
          iOptions: parseIosOptions(args?["ios"], storage.iOptions)!,
          lOptions: storage.lOptions,
          wOptions: parseWindowsOptions(args?["windows"], storage.wOptions)!,
          webOptions: parseWebOptions(args?["web"], storage.webOptions)!,
          mOptions: parseMacOptions(args?["macos"], storage.mOptions as MacOsOptions)!,
        );
      case "clear":
        return await storage.deleteAll(
          aOptions: parseAndroidOptions(args?["android"], storage.aOptions)!,
          iOptions: parseIosOptions(args?["ios"], storage.iOptions)!,
          lOptions: storage.lOptions,
          wOptions: parseWindowsOptions(args?["windows"], storage.wOptions)!,
          webOptions: parseWebOptions(args?["web"], storage.webOptions)!,
          mOptions: parseMacOptions(args?["macos"], storage.mOptions as MacOsOptions)!,
        );
      default:
        throw Exception("Unknown SecureStorage method: $name");
    }
  }

  @override
  void dispose() {
    debugPrint("SecureStorageService(${control.id}).dispose()");
    control.removeInvokeMethodListener(_invokeMethod);
    super.dispose();
  }
}
